tiledb/api/src/array/dimension/mod.rs

```diff
@@ -329,36 +359,54 @@ impl<'ctx> Factory<'ctx> for DimensionData {
     type Item = Dimension<'ctx>;
 
     fn create(&self, context: &'ctx Context) -> TileDBResult<Self::Item> {
-        let mut b = fn_typed!(self.datatype, DT, {
-            let d0 = serde_json::from_value::<DT>(self.domain[0].clone())
-                .map_err(|e| {
-                    Error::Deserialization(
-                        format!("dimension '{}' lower bound", self.name),
-                        anyhow!(e),
-                    )
-                })?;
-            let d1 = serde_json::from_value::<DT>(self.domain[1].clone())
-                .map_err(|e| {
-                    Error::Deserialization(
-                        format!("dimension '{}' upper bound", self.name),
-                        anyhow!(e),
-                    )
-                })?;
-            let extent = serde_json::from_value::<DT>(self.extent.clone())
-                .map_err(|e| {
-                    Error::Deserialization(
-                        format!("dimension '{}' extent", self.name),
-                        anyhow!(e),
-                    )
-                })?;
-            Builder::new::<DT>(
-                context,
-                &self.name,
-                self.datatype,
-                &[d0, d1],
-                &extent,
-            )
-        })?;
+        let mut b = if self.datatype == Datatype::StringAscii {
+            Builder::new_string(context, &self.name, self.datatype)?
+        } else {
+            if self.domain.is_none() {
+                return Err(Error::InvalidArgument(anyhow!(
```

Comment:
I'm not 100% actually. Though looking back, this is all wrong. The way to do this is to remove all of the logic here and go back to relying on core to tell us what's right and wrong. I'll rewrite this so that I can create a builder from a RawDimension and then add the FFI calls to just set what the user presented in the data. That coupled with your idea for the DimensionType enum means that most of the time it'll not get messed up on our end.

Reply:
To be clear, I mean specifically changing the DimensionData domain and extent fields from serde_json values into strongly typed Range values.

Reply:
I replied to the wrong comment... le sigh.

Reply:
Fixed in commit: `Fix Factor implementation for DimensionData`